libraries/AP_Generator/AP_Generator_RichenPower.cpp

```diff
@@ -185,45 +225,67 @@ void AP_Generator_RichenPower::update_runstate()
     static const uint16_t SERVO_PWM_IDLE = 1500;
     static const uint16_t SERVO_PWM_RUN = 1900;
 
-    switch (runstate) {
+    // if the vehicle crashes then we assume the pilot wants to stop
+    // the motor.  This is done as a once-off when the crash is
+    // detected to allow the operator to rearm the vehicle, or we end
+    // up in a catch-22 situation where we force the stop state on the
+    // generator so they can't arm and can't start the generator
+    // because the vehicle is crashed.
+    if (AP::vehicle()->is_crashed()) {
+        if (!vehicle_was_crashed) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Crash; stopping generator");
+            pilot_desired_runstate = RunState::STOP;
+            vehicle_was_crashed = true;
+        }
+    } else {
+        vehicle_was_crashed = false;
+    }
+
+    if (commanded_runstate != pilot_desired_runstate &&
+        !AP::arming().is_armed()) {
+        // consider changing the commanded runstate to the pilot
+        // desired runstate:
+        commanded_runstate = RunState::IDLE;
+        switch (pilot_desired_runstate) {
+        case RunState::STOP:
+            // The H2 will keep the motor running for ~20 seconds for cool-down
+            // we must go via the idle state or the H2 disallows moving to stop!
+            if (time_in_idle_state_ms() > 1000) {
+                commanded_runstate = pilot_desired_runstate;
+            }
+            break;
+        case RunState::IDLE:
+            // can always switch to idle
+            commanded_runstate = pilot_desired_runstate;
+            break;
+        case RunState::RUN:
+            // must have idled for a while before moving to run:
+            if (generator_ok_to_run()) {
+                commanded_runstate = pilot_desired_runstate;
+            }
+            break;
+        }
+    }
+
+    switch (commanded_runstate) {
     case RunState::STOP:
-        _servo_channel->set_output_pwm(SERVO_PWM_STOP);
+        SRV_Channels::set_output_pwm(SRV_Channel::k_richenpower_control, SERVO_PWM_STOP);
         break;
     case RunState::IDLE:
-        _servo_channel->set_output_pwm(SERVO_PWM_IDLE);
+        SRV_Channels::set_output_pwm(SRV_Channel::k_richenpower_control, SERVO_PWM_IDLE);
         break;
     case RunState::RUN:
-        _servo_channel->set_output_pwm(SERVO_PWM_RUN);
+        SRV_Channels::set_output_pwm(SRV_Channel::k_richenpower_control, SERVO_PWM_RUN);
         break;
     }
 }
 
-void AP_Generator_RichenPower::update_servo_channel(void)
+void AP_Generator_RichenPower::Log_Write()
 {
-    const uint32_t now = AP_HAL::millis();
-    if (now - _last_servo_channel_check < 1000) {
+#define MASK_LOG_ANY                    0xFFFF
```

Comment:
I'm surprised we have this definition hidden away here.  normally we put them at the top of the .cpp file.. not a blocker but..